lib/vm/src/instance/mod.rs

```diff
@@ -404,95 +366,38 @@ impl Instance {
-            match self.module.local_func_index(*index) {
-                Some(local_index) => {
-                    let address = self.functions[local_index].0 as u64;
-                    let size = self.function_lengths[local_index];
-                    result.push(NamedFunction {
-                        name: name.clone(),
-                        address,
-                        size,
-                    })
-                }
-                None => {}
-            }
-        }
-        for (name, index) in &self.module.exports {
-            match index {
-                ExportIndex::Function(index) => match self.module.local_func_index(*index) {
-                    Some(local_index) => {
-                        let address = self.functions[local_index].0 as u64;
-                        let size = self.function_lengths[local_index];
-                        result.push(NamedFunction {
-                            name: name.clone(),
-                            address,
-                            size,
-                        })
-                    }
-                    None => {}
-                },
-                _ => {}
-            }
-        }
-        result
+        unsafe { self.vmctx_plus_offset(self.offsets().vmctx_stack_limit_begin()) }
     }
 
     /// Invoke the WebAssembly start function of the instance, if one is present.
-    fn invoke_start_function(&self, trap_handler: &dyn TrapHandler) -> Result<(), Trap> {
-        let start_index = match self.module.start_function {
+    fn invoke_start_function(&self) -> Result<(), Trap> {
+        let start_index = match self.artifact.start_function() {
             Some(idx) => idx,
             None => return Ok(()),
         };
-
-        let (callee_address, callee_vmctx) = match self.module.local_func_index(start_index) {
-            Some(local_index) => {
-                let body = self
-                    .functions
-                    .get(local_index)
-                    .expect("function index is out of bounds")
-                    .0;
-                (
-                    body as *const _,
-                    VMFunctionEnvironment {
-                        vmctx: self.vmctx_ptr(),
-                    },
-                )
-            }
-            None => {
-                assert_lt!(start_index.index(), self.module.num_imported_functions);
-                let import = self.imported_function(start_index);
-                (import.body, import.environment)
-            }
-        };
-
+        let start_funcref = self.funcrefs[start_index];
         // Make the call.
         self.on_call();
         unsafe {
-            catch_traps(trap_handler, || {
+            catch_traps(|| {
                 mem::transmute::<*const VMFunctionBody, unsafe extern "C" fn(VMFunctionEnvironment)>(
-                    callee_address,
-                )(callee_vmctx)
+                    start_funcref.func_ptr,
+                )(start_funcref.vmctx)
```

Comment:
TODO: verify if we need to validate the signature here.

Reply:
Start function does not have any parameters or returns as validated by the wasm validation.